encodings/runend/src/array.rs

```diff
@@ -237,21 +240,115 @@ impl StatisticsVTable<RunEndArray> for RunEndEncoding {
+                    let bools = array.values().into_bool()?.boolean_buffer();
+
+                    let true_count = match array.values().logical_validity() {
+                        LogicalValidity::AllValid(_) => {
+                            match_each_unsigned_integer_ptype!(ends.ptype(), |$P| {
+                                let mut begin = array.offset() as $P;
+                                ends
+                                    .as_slice::<$P>()
+                                    .iter()
+                                    .enumerate()
+                                    .map(|(index, end)| {
+                                        let len = *end - begin;
+                                        begin = *end;
+                                        (len as u64) * (bools.value(index as usize) as u64)
+                                    })
+                                    .sum()
+                            })
+                        }
+                        LogicalValidity::AllInvalid(_) => 0,
+                        LogicalValidity::Array(is_valid) => {
+                            let is_valid = is_valid.into_bool()?.boolean_buffer();
+                            let mut is_valid = is_valid.set_indices();
+                            match is_valid.next() {
+                                None => array.len() as u64,
+                                Some(valid_index) => {
+                                    let offsetted_len = (array.len() + array.offset()) as u64;
+                                    let mut true_count: u64 = array.len() as u64;
+                                    match_each_unsigned_integer_ptype!(ends.ptype(), |$P| {
+                                        let ends = ends.as_slice::<$P>();
+                                        let begin = if valid_index == 0 {
+                                            0
+                                        } else {
+                                            ends[valid_index - 1]
+                                        };
+
+                                        true_count += bools.value(valid_index as usize) as u64 * (cmp::min(ends[valid_index] as u64, offsetted_len) - begin as u64);
+
+                                        for valid_index in is_valid {
+                                            true_count += bools.value(valid_index as usize) as u64 * (cmp::min(ends[valid_index] as u64, offsetted_len) - ends[valid_index - 1] as u64);
+                                        }
+
+                                        true_count
+                                    })
+                                }
+                            }
+                        }
+                    };
+
+                    Some(Scalar::from(true_count))
+                }
+                DType::Primitive(..) => None,
+                dtype => vortex_bail!("invalid dtype: {}", dtype),
+            },
```

Comment:
Can we move this to a separate function?

Reply:
Done.